### PR TITLE
ContainerWait on remove: don't stuck on rm fail

### PR DESCRIPTION
**Pull request by kolyshkin**
Currently, if a container removal has failed for some reason, any client waiting for removal (e.g. `docker run --rm`) is stuck, waiting for removal to succeed while it has failed already. For more details and the reproducer, please check https://github.com/moby/moby/issues/34945

This commit addresses that by allowing `ContainerWait()` with `container.WaitCondition == "removed"` argument to return an error in case of removal failure. The `ContainerWaitOKBody` response returned to a client is amended with an error message, and the `Client.ContainerWait()` is modified to return the error, if any, to the client.

Note that this feature is only available for API version >= 1.34. The old behavior (i.e. do not return from wait on removal error) is being kept for current and older clients, except for the added empty `Error` string field in returned JSON, which to my best knowledge should not cause any compatibility problems.

Now, docker-cli would need a separate commit to bump the API to 1.34 and to show an error returned, if any. The current version of docker-cli will work as is.

* [v2: recreate the waitRemove channel after closing]
* [v3: document new api; keep legacy behavior for older clients]

### CODE REVIEW

**Comment by tiborvass** [comment]
Lol, I just saw this PR, FYI I also opened #35001 may or may not be related.

File: container/state.go
Comment (jlhawn) [inline container/state.go]: Is it necessary to clear the error?

File: container/state.go
Comment (jlhawn) [inline container/state.go]: A problem still exists where another API request could come along after this to try to remove the container again. If it succeeds and `SetRemoved()` is called then it will panic because we've closed this channel again which was already closed.

It should be okay to just create a new `s.waitRemove` channel after closing the original one.

File: container/state.go
Comment (kolyshkin) [reply container/state.go]: The reason to clear it here is, in case the container that failed to start is being removed, this error might not be empty and therefore the client will get a false positive.

Comment (kolyshkin) [reply container/state.go]: Alternatively, we can add another field, like `RemovalError` (to distinguish it from `StartError` which is currently named `Error`).

File: container/state.go
Comment (kolyshkin) [reply container/state.go]: Sure. I have updated the patch to reopen the channel

**Comment by kolyshkin** [comment]
> Lol, I just saw this PR, FYI I also opened #35001. Probably unrelated.

@tiborvass yeah, we both were that close to that lucky number, 350 x :100:

**Comment by vieux** [comment]
ping @tonistiigi @jlhawn

**Review by jlhawn — APPROVED** [review]

**Comment by kolyshkin** [comment]
Note to self: need to test a combo of old daemon and new client (that has `ContainerWaitOKBody.Error` field).

**Comment by kolyshkin** [comment]
Anything else I can do to move this forward? To me it seems quite a critical issue.

**Comment by cpuguy83** [comment]
I don’t know, I’m not sure this is right. It’s a breaking change for sure. The existing behavior is correct in that the container is not removed.

The specific client hang issue could be fixed by after some interval checking on the status (ie, is it in a “Dead” state.

**Comment by kolyshkin** [comment]
> I don’t know, I’m not sure this is right. It’s a breaking change for sure. 

I thought about it, and tried all combinations of old/new client/daemon and it works either same or better. Can you think of any other client that can break so I can check it? Do you mean that with this change, a client can no longer assume the container is removed? This can be fixed a client (check the container status after getting the response, or just check for `response.Error == ""`).

To me, the current API of not reporting a removal error is sort of useless/broken. If we are interested in "container was removed" event, we are most probably also interested in the "container removal failure", too.

In case we want to be extra careful about API, we can deprecate this one and add another one, which works like the one in this PR.

> The specific client hang issue could be fixed by after some interval checking
> on the status (ie, is it in a “Dead” state.

Surely, but this would look like a kludge, a work around broken API. Current way is simple and elegant. Polling for status is not.

**Comment by cpuguy83** [comment]
The issue is more about caller's expectation when the API returns.
We should at least make sure that older API versions keep the old behavior.

**Comment by kolyshkin** [comment]
> We should at least make sure that older API versions keep the old behavior.

I have updated the patch to enable old-style behavior for API version < 1.34. Tested with unmodified docker-cli (which hangs on removal error as it was before), as well as patched client which works as expected (prints an error).

@cpuguy83 PTAL

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: api/types/container/container_wait.go
Comment (kolyshkin) [reply api/types/container/container_wait.go]: Updated as per your recommendation

File: api/server/router/container/container_routes.go
Comment (kolyshkin) [reply api/server/router/container/container_routes.go]: I initially wrote it as `var waitError string` but later changed to `waitError := ""`, as it looks a bit more explicit to me (in terms of what a default value is).

Anyway, this is a struct now.

**Comment by kolyshkin** [comment]
@cpuguy83 I made changes based on your last review, PTAL once time permits.

File: api/server/router/container/container_routes.go
Comment (jlhawn) [inline api/server/router/container/container_routes.go]: Instead of continuing to wait, I think it would be better to just return the error as an internal error. This way, older clients will still unblock but handle it differently and return the error.

Comment (jlhawn) [reply api/server/router/container/container_routes.go]: I mean, this part of the API (error during wait) was unspecified in those older versions but it's better to unblock the client with an error than keep it blocking indefinitely while unaware that an error ever happened.

Comment (cpuguy83) [reply api/server/router/container/container_routes.go]: But the client wouldn't get the error if its an older client and most likely expecting when unblocked that it can perform some other task that it may not be able to due to that error.
I just think this would be a breaking change for older clients that will cause issues in hard to discover ways.
A blocked client is much easier to debug and even potentially work-around.

Comment (jlhawn) [reply api/server/router/container/container_routes.go]: you're right 👍  @kolyshkin please disregard my previous comments.

Comment (kolyshkin) [reply api/server/router/container/container_routes.go]: I agree with @jlhawn, I don't like the fact that we need to update both the daemon and the client for the bug to go away. The problem is, since api version 1.30 we return http 200 OK right away, and so there is no way to return any error later.

Hmm, now when I think about it, there is a way! We can return an error by closing the connection (same thing is happening in case of daemon shutdown), meaning we failed to wait for container removal. This way, docker-cli will be unstuck and print an error ("connection unexpectedly closed" or smth), and in general no client can think that the container is removed as JSON with StatusCode is not received. This is a kludge, too, but it makes sense.

@cpuguy83 what do you think? Oh and by the way, can you point me out to any other client than docker-cli (I'm a bit new to this and don't know of any).

**Comment by kolyshkin** [comment]
OK, I now have three working versions of this patch, all implementing slightly different behavior for older clients in case of removal failure. They are:

1. Same as for new clients (i.e. receive an error which they will ignore as they are not aware).
2. Same as before this patch (i.e. clients keep waiting forever).
3. Same as when daemon exits (i.e. we close the connection), telling the client the wait operation has failed.

Once again, the above are just variations of how to handle old clients when removal has failed. New clients will receive an error as described in the commit message, and can handle it.

We need to choose one. Number 1 is surely an incompatibility in the API. Number 2 is keeping the old (bad) API as is -- the problem is clients are stuck in waiting for something that might never happen. Number 3 is somewhat ugly way to report an error while keeping the API compatible.

Please share your opinion. As it fixes the client stuck I'd love this to move forward.

**Comment by jlhawn** [comment]
I personally like option 3 the best because it at least forces an error on the client, but I-Am-Not-A-Maintainer.

**Comment by cpuguy83** [comment]
Did we test that the docker CLI would return a non-zero exit status with option 3?

**Comment by kolyshkin** [comment]
> Did we test that the docker CLI would return a non-zero exit status with option 3?

I tested all three variations:

## 1. docker-cli from 17.05-ce (client API 1.29 == using legacy wait)

It does not exit with non-zero in case engine is closing the connection. I tested it just by stopping a daemon while having a few `docker run` processes stuck in wait, and got this:

```
ERRO[0043] error getting events from daemon: unexpected EOF 
ERRO[0043] error getting events from daemon: unexpected EOF 
ERRO[0043] error getting events from daemon: unexpected EOF 
EXIT CODE 0
EXIT CODE 0
EXIT CODE 0
```
So, we have a bug in legacy wait code, but it's not the subject of this PR (although it made me think looking into engine legacy wait implementation, as it suffers from the same issue and I didn't realize it).

## 2. docker-cli from 17.06-ce (client API 1.30)

Got this:
```
ERRO[0010] error waiting for container: EOF             
EXIT 125
```

IMHO it's pretty good (and way better than just being stuck).

## 3. docker-cli with my patch (API 1.34)

```
ERRO[0004] Error waiting for container: driver "overlay" failed to remove root filesystem for 5d8a7e9ae4158a3dfce7bed45e9ad1e39fedd8f43f8c94307eed39ea1fbd696b: remove /var/lib/docker/overlay/cb3f2f712b65531bec29d82aa53cf100c01aa138c35335d73539e61dd6397dae/merged: device or resource busy 
EXIT 125
```

**Comment by kolyshkin** [comment]
TL;DR: option number 3 works as intended; need to see the legacyWait, too (but it's a separate issue)

**Comment by jlhawn** [comment]
Docker versions 1.13 through 17.05 don’t use the container wait API for `docker run ...` anyway. It used the Events API as indicated by the error messages.

**Comment by cpuguy83** [comment]
Three works for me then.

**Comment by kolyshkin** [comment]
OK, I have updated the patch to use option 3.

@thaJeztah can you PTAL if API version and docs are fine?

**Comment by kolyshkin** [comment]
So, it seems we have a consensus and I am not aware of anything that prevents this from being merged.

@cpuguy83 I made changes based on your last review two weeks ago, PTAL once time permits.

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Review by jlhawn — APPROVED** [review]
My last review is pretty old, but still LGTM

File: api/swagger.yaml
Comment (thaJeztah) [inline api/swagger.yaml]: Should this use the standard error response type? Similar to https://github.com/kolyshkin/moby/blob/f963500c544daa3c158c0ca3d2985295c875cb6b/api/swagger.yaml#L4334-L4337

Comment (thaJeztah) [reply api/swagger.yaml]: Hm, guess we can't because of the existing `StatusCode` in the response 😞

Comment (kolyshkin) [reply api/swagger.yaml]: AFAIR I tried using `ErrorDetail` and I could not, as it creates a circular package dependency :scream:

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
